keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md

```diff
@@ -0,0 +1,478 @@
+Burstable pod that mounts hugetlbfs and consumes 2GB of 1GB hugepages for packet
+buffer pools. Later, I deploy a Guaranteed pod also requesting 2GB of 1GB hugepages.
+
+**Current behavior**: The Guaranteed pod is admitted (Memory Manager shows
+hugepages as available) but fails at container startup when DPDK tries to allocate
+hugepages that are already consumed.
+
+**Desired behavior**: The Guaranteed pod admission fails immediately with a clear
+error indicating insufficient free hugepages, allowing the scheduler to try
+another node or the administrator to take corrective action.
+
+#### Story 2: Database Workload with Hugepages
+
+As a database administrator, I run PostgreSQL with hugepages enabled for shared
+buffers. If an external monitoring agent or debugging tool temporarily consumes
+hugepages, subsequent Guaranteed pods requesting hugepages should not be admitted
+until hugepages are freed.
+
+**Current behavior**: Pods are admitted based on Memory Manager tracking and fail
+at runtime.
+
+**Desired behavior**: Pods are rejected at admission with informative errors.
+
+### Notes/Constraints/Caveats
+
+- **Race condition window**: A small window exists between verification and actual
+  container startup where hugepages could be consumed. This is inherent to any
+  admission-time check but significantly reduces the failure window compared to
+  no verification.
+
+- **sysfs dependency**: The feature depends on reading from sysfs. If sysfs is
+  unavailable or the free_hugepages file cannot be read, the feature gracefully
+  degrades to current behavior (no verification).
+
+- **Per-NUMA verification**: Verification is performed per-NUMA node, consistent
+  with the Memory Manager's NUMA-aware design.
+
+### Risks and Mitigations
+
+| Risk | Mitigation |
+|------|------------|
+| sysfs reads add latency to admission | Minimal impact: single file read per hugepage size per NUMA node |
+| False rejections due to transient consumption | Acceptable: better to reject than admit and fail at runtime |
+| sysfs unavailable in some environments | Graceful degradation: skip verification if sysfs unreadable |
+
+## Design Details
+
+### Implementation Overview
+
+The implementation consists of two parts:
+
+1. **cadvisor**: Add `FreePages *uint64` field to `HugePagesInfo` struct, populated
+   from sysfs. Uses pointer with `omitempty` to distinguish between "0 free" and
+   "data unavailable".
+
+2. **kubelet Memory Manager**: Add `verifyOSHugepagesAvailability()` function
+   called during `Allocate()` that compares requested hugepages against OS-reported
+   free hugepages from cadvisor's machine info.
+
+### cadvisor Changes
+
+```go
+type HugePagesInfo struct {
+    // huge page size (in kB)
+    PageSize uint64 `json:"page_size"`
+    // number of huge pages
+    NumPages uint64 `json:"num_pages"`
+    // number of free huge pages (nil if unavailable)
+    FreePages *uint64 `json:"free_pages,omitempty"`
```

Comment:
cadvisor maintainers will have the last word of course, but I'd not use a pointer here. Or, from the other angle: why do we need a pointer? I can't recall a platform or a kernel on which free hugepages reporting is not available.